### ISSUE TITLE
High CPU load using the Akka.IO TCP server

### ISSUE THREAD

**Issue by evertmulder** [akkadotnet/akka.net#1225]
The following simple program:
namespace Proto1.Client
{
    using Akka.Actor;
    using Akka.IO;
    using System;

    class Program
    {
        static void Main(string[] args)
        {
            var system = ActorSystem.Create("proto1");
            var manager = system.Tcp(); //this causes 100% CPU
            Console.ReadLine();
        }
    }
}

On my machine it uses around 25% CPU. Starting the console app 4 times gives me 100% CPU utilization:

![2015-08-08 23_45_53-performance monitor](https://cloud.githubusercontent.com/assets/61044/9152416/ded6cd62-3e27-11e5-85de-3c09f86dfeb5.png)

I am using i7 cpu:
![2015-08-08 23_50_42-task manager](https://cloud.githubusercontent.com/assets/61044/9152436/85b87572-3e28-11e5-861c-a2cb11f60683.png)

Using nighly build Akka 1.0.4.64-beta

**Comment by rogeralsing** [comment]
I'm seeing this too, it looks like an entire core is hogged. 
dotTrace points me to `SingleThreadExecutionContext.Execute`

**Comment by rogeralsing** [comment]
Yepp, commenting out the following code makes the issue go away.

```csharp
private readonly BlockingCollection<Action> _queue = new BlockingCollection<Action>();

        public SingleThreadExecutionContext()
        {
//            Task.Factory.StartNew(() =>
//            {
//                foreach (var action in _queue.GetConsumingEnumerable())
//                    action();
//            }, TaskCreationOptions.LongRunning);
        }

        public void Execute(Action action)
        {
            try
            {
                if (!_queue.IsAddingCompleted)
                    _queue.Add(action);
            }
            catch 
            {
                //ignore adding completed
            }
        }
```

**Comment by rogeralsing** [comment]
cc @willieferguson

**Comment by chris-ray** [comment]
+1. Adding a thread sleep in Select() in selectionhandler.cs alleviates the load, indicating some kind of recursion or infinite loop

**Comment by rogeralsing** [comment]
Ok the `Select` method reschedules itself using the `SingleThreadExecutionContext´, so thats the loop.
The JVM impl also does this using the JVM ExecutionContext. so far so good.
The `SingleThreadExecutionContext` seems to be inefficient due to the blockingcollection.

Some questions:

* Does the Select method really need to run in such a tight interval? 
* Do we really need it to run on a completely dedicated thread?

Can't we just run it on the standard thread pool and possibly increase the interval that the Select is scheduled?
(I'm lacking background on this)

**Comment by willieferguson** [comment]
The Select need to run on the same thread - it was initially scheduled on the thread pool, but this caused issues - which is also mentioned in Akka JVM.

I see in Aeron they use an IdleStrategy which does Thread.Yield for a couple of nano-seconds.

I will try this and measure performance & CPU load.

**Comment by willieferguson** [comment]
@evertmulder - please confirm that the PR resolve your issue. I couldn't replicate your high CPU utilization, but reduced load from 10-15% to 0% in my tests.

More details on the issue in #1236.

**Comment by evertmulder** [comment]
Tried the new code. This issue still exists and is not fixed. See comment in 4af2cfb

**Comment by willieferguson** [comment]
Your right - I tested with an TCP listener. There is no point going into the select loop when there are no interested channels.  Will send another PR.

**Comment by evertmulder** [comment]
Just tested the latest dev branch. This bug is fixed. The high CPU load, when no client is connected, is solved.

**Comment by sergiorykov** [comment]
100% CPU load on mono is still there.

Using mono release 4.2.3.4 1mks for Socket.Select timeout causes 100% CPU load. 
I run basic TCP server without connections on Windows and Debian - everywhere 100% CPU (on core).
`
Context.System.Tcp()
       .Tell(new Tcp.Bind(Self, endpoint));
`

I've not figured out why it is happenning. But I've found appropriate value of timeout when CPU load moved to 0.5%. It's around 1000mks. Where is analog of 2ns on JVM but for mono (see comment above)?
`
Socket.Select(readable, writeable, null, 1000);
`

**Should we reopen issue?** 

Actually, in my dirty fix - see attached commit above - I've added additional thread sleep when nobody's connected and nothing has to be sent:
`
if (readable.Count + writeable.Count == 0)
{
        Thread.Sleep(TimeSpan.FromMilliseconds(10));
}
`
In my conditions it's appropriate to miss new connection for 11ms, but is's a failure to has 98.8% CPU on idle tcp server on mono.
